src/tsolvers/liasolver/Matrix.cpp

```diff
@@ -0,0 +1,1053 @@
+                // We add to line l, x times line curcol to check
+                addmul_row(R1, curcol, l, x);
+                submul_column(Ri1, l, curcol, x);
+            }
+            curcol++;
+            if (curcol > U1m.nCols())
+                break;
+        }
+    }
+    curcol--;
+    dim1 = curcol;
+}
+
+
+
+/// Here is the end of computation of Hermite Normal Form ///
+
+
+///compute hnf from pdf
+
+void
+LAMatrixStore::compute_hnf(MId &H, const MId A){//, int &dim) {//maybe no need for dim and p
+
+    LAMatrix &Hm = operator[](H);
+    LAMatrix &Am = operator[](A);
+
+    if (H != A)
+        setMatrix(H, A);
+
+    int k=1, i0;
+    opensmt::Real b, t, q, f, x, y;
+
+    //int p;
+    //for (p = 1; (p <= Am.nCols()) && (p <= Am.nRows()); p++) {
+
+    for (int j = 1; j <= Hm.nCols() && k<=Hm.nRows(); j++) {
+        //k = 1;
+        // Choose Pivot
+        i0 = least_in_col(H, k, j); //i0 should be equal to the number of the row that min val resides
+
+        printf("leastrowvalue %d\n", i0);
+
+        if (i0 > k) {
+            //Exchange row Ak with row Ai0;
+            swap_rows(H, k, i0);
+
+        }
+        if (MM(H, k, j) < 0) {
+            //Ak ← −Ak;
+            neg_row(H, k);
+            //std::cout << " first for loop finished" << "\n";
+        }
+        // Reduce Rows
+        //b ← ak, j;
+        b = MM(H, k,j); //error with k, as MM(mid, row, col) and row maybe 3 but col 4, and in this case k = 4, which breaks assersion
+
+        std::cout << " b equal to " << b << "\n";
+
+        if (b == 0)
+            continue;
+        else {
+            for (int i = k + 1; i <= Hm.nRows(); i++) {
+                //q ← ⌊ai, j / b⌉; //nearest integer or exact div? i checked exact division normal div
+                f = MM(H, i, j) / b; //Note: rounding to nearest integer does not work
+                std::cout << " q1 equal to " << q << "\n";
+                if(!(f.isInteger())){
+                    //get separate den of f (call it x) and cislitel of f (call it y)
+                    x = f.get_den();
+                    y = f.get_num();
+
+                    for (int j = 1; j <= Hm.nCols(); j++) {
+
+                        MM(H, i, j) = x * MM(H, i, j) - y * MM(H, k, j);
+                    }
+
+                } else {
+
+                    //Ai ← Ai − q × Ak;
+                    //MM(H, i, j) -= q * MM(H, k, j);
+                    submul_row(H, i, k, f);
+                }
+            }
+
+            for (int i = k - 1; i > 0; i--) {
+                // q ← ⌊ai,j/b⌋; q = flooor of the division 1/6 and 5.6 shoud give approx 0, if in neg then -1
+                t = MM(H, i, j) / b; //correct this with floor
+                q = t.floor();
+                //Ai ← Ai − q × Ak;
+                //MM(H, i, j) -= q * MM(H, k, j);
+                submul_row(H, i, k, q);
+            }
+        }
+        k = k + 1;
+
+    }
+}
+
+///end of compute hnf from pdf
+
+/*
```

Comment:
Is this code that is commented out needed?

Reply:
For now I'd keep it there.  I'll postpone the decision a little which HNF implementation is the best.  Mostly because I want to be realistic, not because I like it there.